src/ai/backend/manager/repositories/user_resource_policy/db_source/db_source.py

```diff
@@ -0,0 +1,115 @@
+        ),
+        RetryPolicy(
+            RetryArgs(
+                max_retries=5,
+                retry_delay=0.1,
+                backoff_strategy=BackoffStrategy.FIXED,
+                non_retryable_exceptions=(BackendAIError,),
+            )
+        ),
+    ]
+)
+
+
+class UserResourcePolicyDBSource:
+    """
+    Database source for user resource policy operations.
+    Handles all database operations for user resource policies.
+    """
+
+    _db: ExtendedAsyncSAEngine
+
+    def __init__(self, db: ExtendedAsyncSAEngine) -> None:
+        self._db = db
+
+    @user_resource_policy_db_source_resilience.apply()
+    async def create(self, creator: UserResourcePolicyCreator) -> UserResourcePolicyData:
+        """Creates a new user resource policy."""
+        async with self._db.begin_session() as db_sess:
+            db_row = UserResourcePolicyRow.from_creator(creator)
+            db_sess.add(db_row)
+            await db_sess.flush()
+            return db_row.to_dataclass()
+
+    @user_resource_policy_db_source_resilience.apply()
+    async def get_by_name(self, name: str) -> UserResourcePolicyData:
+        """Retrieves a user resource policy by name."""
+        async with self._db.begin_readonly_session() as db_sess:
+            query = sa.select(UserResourcePolicyRow).where(UserResourcePolicyRow.name == name)
+            row = await db_sess.scalar(query)
+            if row is None:
+                raise UserResourcePolicyNotFound(
+                    f"User resource policy with name {name} not found."
+                )
+            return row.to_dataclass()
+
+    @user_resource_policy_db_source_resilience.apply()
+    async def update(
+        self, name: str, modifier: UserResourcePolicyModifier
+    ) -> UserResourcePolicyData:
+        """Updates an existing user resource policy."""
+        async with self._db.begin_session() as db_sess:
+            # Check if the policy exists first
+            check_query = sa.select(UserResourcePolicyRow).where(UserResourcePolicyRow.name == name)
+            existing_row: Optional[UserResourcePolicyRow] = await db_sess.scalar(check_query)
+            if existing_row is None:
+                raise UserResourcePolicyNotFound(
+                    f"User resource policy with name {name} not found."
+                )
+
+            fields = modifier.fields_to_update()
+            update_stmt = (
+                sa.update(UserResourcePolicyRow)
+                .where(UserResourcePolicyRow.name == name)
+                .values(**fields)
+                .returning(UserResourcePolicyRow)
+            )
+            query_stmt = (
+                sa.select(UserResourcePolicyRow)
+                .from_statement(update_stmt)
+                .execution_options(populate_existing=True)
+            )
+            updated_row: Optional[UserResourcePolicyRow] = await db_sess.scalar(query_stmt)
```

Comment:
The query feels unnecessarily complicated.
It seems like using the returning value of the update would work, but is an additional select necessary?